fcrepo/fcrepo
# FCREPO-3541 - Etags reflect membership and other request preferences

**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3541

# What does this Pull Request do?
* Determines last updated timestamp of membership for a resource
* Etags and state tokens now diverge for RDF sources, as etags now reflect
    * State token
    * Requested mimetype
    * Prefer headers that affect the included triples
    * Last updated from containment index
    * Last updated from membership index
* Adds a service for calculating etags
* Removes inclusion of containment last updated from state token calculation, as it is happening at request time now


# How should this be tested?

You can see a few other examples of DirectContainer operations that could be used for testing in:
https://github.com/fcrepo/fcrepo/pull/1781

```
# Create the membership resource
curl -XPUT http://localhost:8080/rest/mem_resc1 -ufedoraAdmin:fedoraAdmin

# Create a DirectContainer using hasMember relation
echo "
<http://localhost:8080/rest/dc_hm> <http://www.w3.org/ns/ldp#membershipResource> <http://localhost:8080/rest/mem_resc1> .
<http://localhost:8080/rest/dc_hm> <http://www.w3.org/ns/ldp#hasMemberRelation> <http://example.com/hasMember> .
" | curl -XPUT http://localhost:8080/rest/dc_hm -ufedoraAdmin:fedoraAdmin -H'Link: <http://www.w3.org/ns/ldp#DirectContainer>;rel="type"' -H"Content-Type: application/n-triples" --data-binary "@-"

# Etag for GET and HEAD requests should be the same

curl -I http://localhost:8080/rest/mem_resc1 -ufedoraAdmin:fedoraAdmin

    ETag: W/"E1F2B178018FD97465BC14AD84826960"
  
curl http://localhost:8080/rest/mem_resc1 -ufedoraAdmin:fedoraAdmin -v

    < ETag: W/"E1F2B178018FD97465BC14AD84826960"

# Check etag without membership, should be the same at this point
curl http://localhost:8080/rest/mem_resc1 -ufedoraAdmin:fedoraAdmin -v -H 'Prefer: return=representation; omit="http://www.w3.org/ns/ldp#PreferMembership"'

    < ETag: W/"E1F2B178018FD97465BC14AD84826960"


# Now add a member
curl -XPUT http://localhost:8080/rest/dc_hm/memb1 -ufedoraAdmin:fedoraAdmin

# Etag in HEAD and GET should have changed from before, but match each other
curl -I http://localhost:8080/rest/mem_resc1 -ufedoraAdmin:fedoraAdmin

    ETag: W/"8C2E7DFF1429384C8D97E0F35A02E49A"
    
curl http://localhost:8080/rest/mem_resc1 -ufedoraAdmin:fedoraAdmin -v

    < ETag: W/"8C2E7DFF1429384C8D97E0F35A02E49A"

# ETag without membership should match from before membership added
curl http://localhost:8080/rest/mem_resc1 -ufedoraAdmin:fedoraAdmin -v -H 'Prefer: return=representation; omit="http://www.w3.org/ns/ldp#PreferMembership"'

    < ETag: W/"E1F2B178018FD97465BC14AD84826960"
    


# Verify that excluding containment impacts etag
curl -XPUT http://localhost:8080/rest/parent_resc -ufedoraAdmin:fedoraAdmin

# Etag of resc before children
curl -I http://localhost:8080/rest/parent_resc -ufedoraAdmin:fedoraAdmin

    ETag: W/"A3CFA72A1C75348BC44B463F7F98BE86"

# Add child
curl -XPUT http://localhost:8080/rest/parent_resc/child_resc -ufedoraAdmin:fedoraAdmin

# Etag of parent should have changed
curl -I http://localhost:8080/rest/parent_resc -ufedoraAdmin:fedoraAdmin

    ETag: W/"B3CC7929EA643EC9BA83E5C32F2ECDE7"

# Excluding containment should match etag before children
curl http://localhost:8080/rest/parent_resc -ufedoraAdmin:fedoraAdmin -v -H 'Prefer: return=representation; omit="http://www.w3.org/ns/ldp#PreferContainment"'

    < ETag: W/"A3CFA72A1C75348BC44B463F7F98BE86"
```
